### ISSUE TITLE
Templates don't rerender when using route groups

### ISSUE THREAD

**Issue by tom-sherman** [vercel/next.js#52422]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.5.0: Thu Jun  8 22:22:20 PDT 2023; root:xnu-8796.121.3~7/RELEASE_ARM64_T6000
    Binaries:
      Node: 18.13.0
      npm: 8.13.2
      Yarn: 1.22.19
      pnpm: 8.6.6
    Relevant Packages:
      next: 13.4.10-canary.0
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 4.9.4
    Next.js Config:
      output: N/A
```

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/tom-sherman/nextjs-template-route-group-bug

### To Reproduce

1. Add a template in the same route as a route group
2. Navigate between pages in the group

### Describe the Bug

Template should rerender. In the linked reproduction, we should see new random numbers rendered to the screen when navigating between A & B using the links.

### Expected Behavior

Template doesn't rerender. Numbers don't change on navigation between A & B

**Comment by Fredkiss3** [comment]
Hello, i tested the code and it seems templates just don't rerender on navigation wether in route groups or not (tested by replacing `(group)` with `/group`). 

And i think, when the docs means they rerender on navigation, it is only in the client.  
As said in the docs, the children of templates are totally replaced (on the client) : 

 > Templates are similar to layouts in that they wrap each child layout or page. Unlike layouts that persist across routes and maintain state, templates create a new instance for each of their children on navigation. This means that when a user navigates between routes that share a template, a new instance of the component is mounted, DOM elements are recreated, state is not preserved, and effects are re-synchronized.

https://nextjs.org/docs/app/building-your-application/routing/pages-and-layouts#templates

I think in the server side, they behave just like templates.

**Comment by Fredkiss3** [comment]
I created an example with the behavior outlined in the docs : https://github.com/Fredkiss3/next-template-correct-behavior

In the example i just made the nested layout & nested templates as client components. You can see when navigating from one route to another the number on the nested template changes everytime, while it stays the same for the nested layout.

**Comment by tom-sherman** [comment]
> And i think, when the docs means they rerender on navigation, it is only in the client.

Server templates do rerender when they're used in a dynamic segment, but they have to be at the same level as the dynamic page.

https://github.com/tom-sherman/nextjs-template-route-group-bug/commit/bfed21ce22342b0b73ff03cfbedb8424776761ac

I suppose it makes sense, the template in the route group case is receiving only the `routeParam` that is knows about and using that as the `key`, it can't know it's child params or routes.

**Comment by tom-sherman** [comment]
> I think in the server side, they behave just like layouts.

Ah, you're right and this makes sense. I guess this is just expected behaviour then

**Comment by Fredkiss3** [comment]
> Server templates do rerender when they're used in a dynamic segment, but they have to be at the same level as the dynamic page.

I think it is the same for layouts, they need to be aware of the param, so if it changes, the layout needs to rerender (on the server) also.

**Comment by tom-sherman** [comment]
It would be nice if server templates were aware of route groups, but I suppose that's a feature request not a bug.

**Comment by maiconsanson** [comment]
> Hello, i tested the code and it seems templates just don't rerender on navigation wether in route groups or not (tested by replacing `(group)` with `/group`).
> 
> And i think, when the docs means they rerender on navigation, it is only in the client. As said in the docs, the children of templates are totally replaced (on the client) :
> 
> > Templates are similar to layouts in that they wrap each child layout or page. Unlike layouts that persist across routes and maintain state, templates create a new instance for each of their children on navigation. This means that when a user navigates between routes that share a template, a new instance of the component is mounted, DOM elements are recreated, state is not preserved, and effects are re-synchronized.
> 
> https://nextjs.org/docs/app/building-your-application/routing/pages-and-layouts#templates
> 
> I think in the server side, they behave just like layouts.

Exactly! But if I want to use a template I would expect the DOM elements to be recreated (especially on the server side, that's the main difference of the Layout). This is a bug or is badly documented.

**Comment by KieronWiltshire** [comment]
This seems like a huge bug. Stupid imo and needs fixing on NextJS 15 before release smh.

**Comment by marcoshernanz** [comment]
Thanks for the report. I retested the reproduction on `13.4.10-canary.0` and current canary. The template inside `(group)` is remounted between `/a` and `/b`, and its DOM node is recreated on every navigation. The `Math.random()` value comes from a Server Component and may be reused, so it is not a reliable remount check. This matches the documented [template behavior](https://nextjs.org/docs/app/api-reference/file-conventions/template#behavior). Closing as expected behavior.
